docs/adr/0010-transition-fxa-from-backbone-to-react.md

```diff
@@ -0,0 +1,151 @@
+  - Whatever new library or framework we choose will require a learning curve for at least part of the team. It may slow the ability to make front-end changes at first.
+
+### Option B - Move to client-side React
+
+- Description:
+  - Use client-side React.
+- Pros
+  - React is backed by a well-known company (Facebook) with a lot of dedicated resources working on the library and it's popular, has a lot of community support, and has ample documentation.
+  - The `fxa-payments-server` is already using React. We can begin thinking about reusing components across the FxA monorepo.
+  - React promotes simplicity, clarity, and predictability. We know that data exists at the top and it flows down to children components, and we know what the data is with prop and type checking. The component then takes the data and subjects it to lifecycle methods that execute in a predictable manner every time. Backbone, on the other hand, creates a JSON object that is exposed to mustache files, but we don't always know what this data is or what its types are. Backbone also oftentimes influences views via updates through a model, multiple models, and/or controllers. It's not as clear when a view or its children will be rerendered or a DOM element updated.
+  - React is fast. It bundles state updates together and figures out the fewest operations it can get away with, giving us optimization wins out of the box.
+  - React promotes declarative programming over imperative. In Backbone, we interact with the DOM directly via jQuery, _telling_ the browser exactly what to do (imperative) but with React, we _describe_ the UI with components and let React and the browser figure out how to implement it (declarative). This can be much less tedious for developers and allows React to handle state changes optimally.
+  - React promotes the idea of composition over inheritance. In Backbone, we extend views for reusability, which can cause difficulty in seeing or realizing what events and other methods or data you are inheriting (the banana, monkey, jungle problem) and overwriting. The intention of components in React is more clear and data is more deliberately passed around.
+  - React offers many integration options.
+  - React has already been audited by our security team.
+- Cons
+  - React changes fairly often. The definition of "knowing React" will continue to grow as new concepts and best practices, like React Hooks or depreciated lifecycle methods, are created or recommended.
```

Comment:
In practice how big have you seen these changes to be? Would we have to do an ADR to potentially upgrade to a new React version because of breaking changes etc?

Reply:
I don't believe so - the largest change I've seen and have read about has been around the React 16 upgrade introducing hooks, which introduced "no breaking changes" (see [this section of the React docs](https://reactjs.org/docs/faq-versioning.html) for more info). When the last team I worked with did this upgrade, we did have to make a couple of minor changes since we were using React a little unconventionally, but it wasn't too much effort. They continue to support "older" ways of doing things, but we could potentially have some console warnings to clean up with upgrades.